### PR TITLE
Modulo 3: Implement Movie List from local JSON and navigation into Details page.

### PR DESCRIPTION
**Pull request by jere53**
This PR implementes a list of Movies obtained from a mocked local source using JSON as described [here](https://docs.google.com/document/d/1nn9M4O7mLzOVZBS8aPtz8BZO3hQJ7FEI/edit?pli=1). Navigation between the Movie list and each Movie Details page can be achieved by tapping the corresponding entry in the Movie List.

A Repository class was created to handle JSON loading.

Since the mocked source doesn't include Image files, the code was modified to use a placeholder in case of an invalid image path.

Here is how the UI looks like when the image paths are invalid:
[with placeholder.webm](https://github.com/jere53/jeremias.salinas/assets/64437798/ea9b244f-bc25-4ed2-a90a-615cf2f5ef48)

And here is how the UI looks like when the images are valid:
[with data.webm](https://github.com/jere53/jeremias.salinas/assets/64437798/586e3f5a-38c8-40b7-b7cf-63f78d1c06c3)

### CODE REVIEW

File: lib/main.dart [hunk 1]
```diff
@@ -1,67 +1,18 @@
 import 'package:flutter/material.dart';
-import '../../../utils/ui_constants.dart';
-import 'custom_widgets.dart';
-import '../../../models/movie.dart';
+import 'package:movie_app_jsalinas/pages/movies_page.dart';
```
Comment (MarcosCGlobant) [inline lib/main.dart:2]: relative import here please

File: lib/models/genres.dart [hunk 2]
```diff
@@ -0,0 +1,10 @@
+class Genre {
+  final int id;
+  final String name;
+
+  const Genre(this.id, this.name,);
+
+  factory Genre.fromJson(Map<String, dynamic> json){
+    return Genre(json['id'], json['name'],);
```
Comment (MarcosCGlobant) [inline lib/models/genres.dart:8]: format code here

File: lib/models/movie.dart [hunk 3]
```diff
@@ -1,3 +1,5 @@
+import 'package:movie_app_jsalinas/models/genres.dart';
```
Comment (MarcosCGlobant) [inline lib/models/movie.dart:1]: relative import here please

File: lib/models/movie.dart [hunk 4]
```diff
@@ -39,5 +53,36 @@ class Movie {
...
+        false,
+        100000,
       );
+
+  factory Movie.fromJson(Map<String, dynamic> json, List<Genre> genres) {
+    List<dynamic> movieGenreIDs = json['genre_ids'] as List<dynamic>;
+    List<String> movieGenres = genres
+        .where((e) => movieGenreIDs.contains(e.id))
+        .map((e) => e.name)
+        .toList();
```
Comment (MarcosCGlobant) [inline lib/models/movie.dart:69]: you could create a factory method fromList to get the genres inside the Genre class, like this

```
static List<Genre>? fromList(List<dynamic>? list) {
    var result = <Genre>[];
    list?.forEach(
      (item) {
        var genre = Genre.fromJson(item);
        if (genre != null) {
          result.add(genre);
        }
      },
    );

    return result;
  }
```

File: lib/pages/movies_page.dart [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+                          );
+                        },
+                        child: buildMovieCard(movie),
+                      ),
+                  ],
+                ),
+              ),
+            );
+          }
+          return const Text('no data');
```
Comment (MarcosCGlobant) [inline lib/pages/movies_page.dart:63]: hardcode into a constant

File: lib/pages/movies_page.dart [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+        });
+  }
+
+  Card buildMovieCard(Movie movie) {
```
Comment (MarcosCGlobant) [inline lib/pages/movies_page.dart:67]: the build method, should be the last on the file, move the bellow methods above it

File: lib/utils/movie_repository.dart [hunk 7]
```diff
@@ -0,0 +1,53 @@
+import 'dart:convert';
+import 'package:flutter/services.dart';
+import 'package:movie_app_jsalinas/models/genres.dart';
+import 'package:movie_app_jsalinas/models/movie.dart';
```
Comment (MarcosCGlobant) [inline lib/utils/movie_repository.dart:4]: relative imports here please

File: lib/utils/movie_repository.dart [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+class MovieRepository {
```
Comment (MarcosCGlobant) [inline lib/utils/movie_repository.dart:6]: this class looks like it could be separated into an Interface and the respective implementation of the Interface, in 2 different files

File: lib/widgets/genre_list.dart [hunk 9]
```diff
@@ -1,17 +1,13 @@
 import 'package:flutter/material.dart';
+import 'package:movie_app_jsalinas/utils/ui_constants.dart';
```
Comment (MarcosCGlobant) [inline lib/widgets/genre_list.dart:2]: relative import here please

File: lib/widgets/movie_poster.dart [hunk 10]
```diff
@@ -21,12 +24,21 @@ class MoviePoster extends StatelessWidget {
...
-            fit: BoxFit.fill,
-          ),
+        ),
+        child: Image.asset(
+          pathToPosterImg,
+          height: height,
+          width: width,
+          fit: BoxFit.cover,
+          errorBuilder:
+              (BuildContext context, Object exception, StackTrace? stackTrace) {
```
Comment (MarcosCGlobant) [inline lib/widgets/movie_poster.dart:34]: trailing comma and format code here please

File: lib/widgets/movie_score.dart [hunk 11]
```diff
@@ -5,16 +5,32 @@ class MovieScore extends StatelessWidget {
...
     fontStyle: FontStyle.italic,
   );
+  static const TextStyle voteCountTextStyle = TextStyle(
+    color: Colors.white,
+    fontSize: 15,
+  );
+
+  String _getFormattedVoteCount() {
+    if (voteCount > 99999) {
+      return '99999+';
```
Comment (MarcosCGlobant) [inline lib/widgets/movie_score.dart:29]: values into constants

File: lib/widgets/movie_score.dart [hunk 12]
```diff
@@ -27,10 +43,26 @@ class MovieScore extends StatelessWidget {
...
+                textAlign: TextAlign.center,
+                movieScore,
+                style: scoreTextStyle,
+              ),
+            ),
+            Flexible(
+              flex: voteCountFlex,
+              child: Text(
+                textAlign: TextAlign.center,
+                '${_getFormattedVoteCount()} votes',
```
Comment (MarcosCGlobant) [inline lib/widgets/movie_score.dart:61]: hardcode value into constant please

**Review by MarcosCGlobant — CHANGES_REQUESTED** [review]

File: lib/models/movie.dart (hunk 4 above)
Comment (jere53) [reply lib/models/movie.dart:69]: Is it best to leave this in the Genre class or use a genresFromList method in Movie? It makes sense for it to be in Genres but when called "Genre.fromList()" sounds like it should return a Genre, not a list of Genres.

Comment (jere53) [reply lib/models/movie.dart:69]: A second question now that I'm re-reading the code...Would it be better if the Genre class had a list with all the possible Genres, and a static "getGenre(id)" method that I could call from the Movie, or should I keep the current approach of passing a List of possible Genres to the Movie, and having the Movie decide which ones it has?

File: lib/models/movie.dart
Comment (matiasilveiro-gl) [inline lib/models/movie.dart:1]: It's a bit late, but it'd be great to have named parameters for the constructor. That'd avoid mistakes when your constructor has multiple parameters like this one.

File: lib/pages/movies_page.dart
Comment (matiasilveiro-gl) [inline lib/pages/movies_page.dart:1]: Make widget builder methods private, like `_buildMovieCard` and so on. Or you can even extract them as separate widgets. As you prefer.

File: lib/pages/movies_page.dart [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+                        },
+                        child: buildMovieCard(movie),
+                      ),
+                  ],
+                ),
+              ),
+            );
+          }
+          return noDataText;
+        });
```
Comment (matiasilveiro-gl) [inline lib/pages/movies_page.dart:162]: Trailing comma

File: lib/pages/movies_page.dart [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+          ),
+        ),
+      ),
+    );
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return FutureBuilder(
+        future: MovieRepository().getMovies(),
```
Comment (matiasilveiro-gl) [inline lib/pages/movies_page.dart:127]: We can add some Flutter best practices here. Please read this:
https://api.flutter.dev/flutter/widgets/FutureBuilder-class.html

> If the future is created at the same time as the FutureBuilder, then every time the FutureBuilder's parent is rebuilt, the asynchronous task will be restarted. A general guideline is to assume that every build method could get called every frame, and to treat omitted calls as an optimization.

TLDR, you could make this widget `Stateful`, declare a `Future<T> movies` at the beginning, init that future in the `initState` method, and use that future in the `FutureBuilder`. Like here:
https://stackoverflow.com/a/68966640

**Review by matiasilveiro-gl — CHANGES_REQUESTED** [review]

File: lib/pages/movies_page.dart (hunk 14 above)
Comment (jere53) [reply lib/pages/movies_page.dart:127]: nice, thats good to know. Fixed!

File: lib/models/movie.dart (hunk 4 above)
Comment (MarcosCGlobant) [reply lib/models/movie.dart]: for the first question, its better to have it on Genre, to keep the related methods according to the class, its not wrong that Genre.fromList returns a list of Genres, better for Movie to receive the info that to transform it.
And for the second, the method sounds better, you could use it inside the fromJson

File: lib/main.dart [hunk 15]
```diff
@@ -1,67 +1,18 @@
...
-                movieOverviewTxt: movie.overview,
-              ),
-            ],
-          ),
-        ),
-      ),
+      initialRoute: MoviesPage.routeName,
+      routes: {
+        MoviesPage.routeName: (context) => const MoviesPage(),
+      },
```
Comment (matiasilveiro-gl) [inline lib/main.dart:15]: To debate: why do you define named routes and then navigate by pushing `MaterialPage`s? It'd be great to standardize a single approach to your app's navigation.

Comment (jere53) [reply lib/main.dart:15]: I initially used named routes because it seemed like the de-facto more flexible choice. When it came to implementing the actual functionality though, I found that a stack modelled it perfectly, and that it was clearer to just say "push the Details page for this movie" without worrying about the route. Because of this I removed Details from the routes but left the Movies. At the moment, with only 2 pages, named routes don't really add anything. However later on if more pages (like a Favorites page) are added and we can reach Details from more than one place, it will be better for them to be named. If it's better to have a single routing scheme through the whole app, then it might be worth it to change Details to use a named route as well. Should I change this now or wait until there are actually more pages?

File: lib/models/movie.dart (hunk 4 above)
Comment (jere53) [reply lib/models/movie.dart]: Sorted, thank you!

**Review by MarcosCGlobant — APPROVED** [review]

**Review by matiasilveiro-gl — APPROVED** [review]
LGTM!

File: lib/models/genres.dart [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+    );
+    return result;
+  }
+
+  static List<Genre> getGenresFromIDs(List<dynamic> idList) {
+    List<Genre> result =
+        validGenres.where((e) => idList.contains(e.id)).toList();
+
+    return result;
+  }
```
Comment (FranciscoGaraventa) [inline lib/models/genres.dart:33]: This could be moved to another file, Genre should only need to know how to return itself not a Genres list, you can create a class like GenreModel that has a parameter list<Genre> genres

File: lib/models/movie.dart [hunk 17]
```diff
@@ -7,37 +9,79 @@ class Movie {
...
-        [
+        releaseDate: '2023-04-05',
+        voteAverage: 7.8,
+        genres: [
           'Animation',
           'Family',
           'Adventure',
           'Fantasy',
           'Comedy',
         ],
```
Comment (FranciscoGaraventa) [inline lib/models/movie.dart:53]: With the new genres data that you have here instead to use the list of string, you can have a list of int, with the id's for the genres

File: lib/pages/movies_page.dart [hunk 18]
```diff
@@ -0,0 +1,176 @@
...
+        ),
+      ),
+    );
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return FutureBuilder(
+      future: movies,
+      builder: (BuildContext context, AsyncSnapshot snapshot) {
```
Comment (FranciscoGaraventa) [inline lib/pages/movies_page.dart:139]: trailing comma

Comment (FranciscoGaraventa) [reply lib/pages/movies_page.dart:139]: And also you can specify on AsyncSnapshot the data that you are expecting like List<Movie> in this case, to avoid cast on line 145

File: lib/utils/movie_repository_from_json.dart [hunk 19]
```diff
@@ -0,0 +1,38 @@
+import 'dart:convert';
+import 'package:flutter/services.dart';
+import '../../../models/genres.dart';
+import '../../../models/movie.dart';
+import '../../../utils/movie_repository.dart';
+
+class MovieRepositoryFromJson implements MovieRepository {
```
Comment (FranciscoGaraventa) [inline lib/utils/movie_repository_from_json.dart:7]: You can rename this, is not necessary to specify that you are getting the data from Json, so the class name could be renamed to MovieRepository and the interface you can renamed to IMovieRepository

File: pubspec.yaml [hunk 20]
```diff
@@ -31,6 +31,7 @@ flutter:
 
   assets:
     - images/
+    - data/
```
Comment (FranciscoGaraventa) [inline pubspec.yaml:34]: Maybe you can renamed to mock, do you think that is more appropiated name?

File: lib/models/genres.dart (hunk 16 above)
Comment (jere53) [reply lib/models/genres.dart:33]: Makes sense, I'm guessing this is the difference between Model and Entity (?

File: lib/pages/movies_page.dart (hunk 18 above)
Comment (jere53) [reply lib/pages/movies_page.dart:139]: good tip, thanks!

File: lib/utils/movie_repository_from_json.dart (hunk 19 above)
Comment (jere53) [reply lib/utils/movie_repository_from_json.dart:7]: Done

File: pubspec.yaml (hunk 20 above)
Comment (jere53) [reply pubspec.yaml:34]: makes sense to me, changed!

File: lib/models/movie.dart (hunk 17 above)
Comment (jere53) [reply lib/models/movie.dart]: ended up making a new MovieModel class that works with IDs and maps them to a Movie (entity) class that references the actual Genre entity

File: lib/models/movie_model.dart [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+        pathToBackdropImg: pathToBackdropImg,
+        pathToPosterImg: pathToPosterImg,
+        adult: adult,
+        originalLanguage: originalLanguage,
+        id: id,
+        popularity: popularity,
+        video: video,
+        voteCount: voteCount);
+  }
+}
```
Comment (FranciscoGaraventa) [inline lib/models/movie_model.dart:103]: Do you really need this method?

File: lib/pages/movies_page.dart [hunk 22]
```diff
@@ -0,0 +1,180 @@
...
+  void initState() {
+    super.initState();
+    movies = IMovieRepository().getMovies();
+  }
+
+  Card _buildMovieCard(Movie movie) {
+    String releaseYear = DateTime.parse(movie.releaseDate).year.toString();
+
+    return Card(
+      color: Colors.blueGrey[800],
```
Comment (FranciscoGaraventa) [inline lib/pages/movies_page.dart:44]: You can change this value to Colors.blueGrey.shade800

File: lib/pages/movies_page.dart [hunk 23]
```diff
@@ -0,0 +1,180 @@
...
+                          fontSize: MoviesPage.titleFontSize,
+                          color: Colors.white,
+                          fontWeight: FontWeight.bold,
+                        ),
+                      ),
+                      Container(height: MoviesPage.textSpacing),
+                      Text(
+                        releaseYear,
+                        style: TextStyle(
+                          color: Colors.grey[300],
```
Comment (FranciscoGaraventa) [inline lib/pages/movies_page.dart:79]: You can change this value to Colors.grey.shade300

File: lib/models/genre.dart [hunk 24]
```diff
@@ -0,0 +1,17 @@
+class Genre {
+  final int id;
+  final String name;
+  static List<Genre> validGenres = [];
```
[the remark this answers is not in the document]
Comment (jere53) [reply lib/models/genre.dart:4]: whoops, fixed it

File: lib/models/movie_model.dart (hunk 21 above)
Comment (jere53) [reply lib/models/movie_model.dart:103]: Hmmm, it is true that I call this method after every call to MovieModel's constructors since I only ever work with the entity. I replaced the MovieModel named constructors with static methods that return a Movie entity. Since we never actually use a MovieModel by itself, I also removed every one of its fields. Now it's just a class with no fields that provides static methods which return Movie entities. Is "MovieModel" still an appropriate name?

File: lib/pages/movies_page.dart (hunk 23 above)
Comment (jere53) [reply lib/pages/movies_page.dart:79]: done. Is there an advantage to this way of doing it, or is it a styling thing?

File: lib/pages/movies_page.dart (hunk 22 above)
Comment (jere53) [reply lib/pages/movies_page.dart:44]: done

File: lib/pages/movies_page.dart (hunk 23 above)
Comment (FranciscoGaraventa) [reply lib/pages/movies_page.dart]: Yes, depends also to the pallete color that has the Theme because if you are trying to get a color that doesn't exist you can have an error like this
`type 'Color' is not a subtype of type 'MaterialColor'`

File: lib/models/movie.dart [hunk 25]
```diff
@@ -7,37 +7,29 @@ class Movie {
   final List<String> genres;
```
Comment (FranciscoGaraventa) [inline lib/models/movie.dart:7]: Change genres to a List<int> instead to List<String>

File: lib/models/movie_model.dart [hunk 26]
```diff
@@ -0,0 +1,59 @@
+import 'movie.dart';
+
+import 'genre_model.dart';
+
+class MovieModel {
```
Comment (FranciscoGaraventa) [inline lib/models/movie_model.dart:5]: This class is not necessary at all, you have the Movie class created and also you have the GenresModel that has a List<Genres>

Comment (jere53) [reply lib/models/movie_model.dart:5]: makes sense, fixed

**Review by FranciscoGaraventa — APPROVED** [review]
LG

File: lib/pages/movies_page.dart [hunk 27]
```diff
@@ -107,7 +106,23 @@ class _MoviesPageState extends State<MoviesPage> {
...
+        (g) => genres.contains(
+          g.id,
+        ),
+      ),
+    );
+    for (Genre genre in transformedGenres) {
+      children.add(_buildGenreCard(genre));
+    }
+    return children;
+  }
```
Comment (Kevin-Guillen) [inline lib/pages/movies_page.dart:123]: i think we can find a more efficient way to build this method using a map

Comment (jere53) [reply lib/pages/movies_page.dart]: done!

**Review by Kevin-Guillen — APPROVED** [review]

File: lib/main.dart (hunk 1 above)
Comment (jere53) [reply lib/main.dart]: solved

File: lib/models/genres.dart (hunk 2 above)
Comment (jere53) [reply lib/models/genres.dart]: solved

File: lib/models/movie.dart (hunk 3 above)
Comment (jere53) [reply lib/models/movie.dart]: solved

File: lib/pages/movies_page.dart (hunk 5 above)
Comment (jere53) [reply lib/pages/movies_page.dart]: solved

File: lib/pages/movies_page.dart (hunk 6 above)
Comment (jere53) [reply lib/pages/movies_page.dart]: solved

File: lib/utils/movie_repository.dart (hunk 7 above)
Comment (jere53) [reply lib/utils/movie_repository.dart]: solved

File: lib/utils/movie_repository.dart (hunk 8 above)
Comment (jere53) [reply lib/utils/movie_repository.dart]: solved

File: lib/widgets/genre_list.dart (hunk 9 above)
Comment (jere53) [reply lib/widgets/genre_list.dart]: solved

File: lib/widgets/movie_poster.dart (hunk 10 above)
Comment (jere53) [reply lib/widgets/movie_poster.dart]: solved

File: lib/widgets/movie_score.dart (hunk 11 above)
Comment (jere53) [reply lib/widgets/movie_score.dart]: solved

File: lib/widgets/movie_score.dart (hunk 12 above)
Comment (jere53) [reply lib/widgets/movie_score.dart]: solved
